1_preprocessing_data/PyBaSiC_Pipelines/scripts/nbconverted/Illumination_Correction.py

```diff
@@ -0,0 +1,224 @@
+
+
+def run_illum_correct(channel_path: pathlib.Path, channel: str, save_path: pathlib.Path, save_channel: str, output_calc: str = 'False'):
+    """ calculates flatfield, darkfield, performs illumination correction on channel images., coverts to 8-bit and saves images into designated folder
+
+    Parameters:
+        channel_path (pathlib.Path): path to directory where image channels are located
+        channel (str): name of channel
+        save_path (pathlib.Path): path to directory where the corrected images will be saved to designated folders
+        channel (str): name of channel with "_Corrected" (or however you want to designate the images) to create folder to save to
+    """
+    # Loads in the variables returned from "load_pybasic_data" function
+    channel_images, image_files = load_pybasic_data(channel_path, channel)
+
+    flatfield, darkfield = pybasic.basic(channel_images, darkfield=True)
+
+    # Optional output that displays the plots for the flatfield and darkfield calculations if set to True (default is False)
+    if output_calc == True:
+        plt.title('Flatfield')
+        plt.imshow(flatfield)
+        plt.colorbar()
+        plt.show()
+        plt.title('Darkfield')
+        plt.imshow(darkfield)
+        plt.colorbar()
+        plt.show()
+
+    # Run PyBaSiC illumination correction function on loaded in images
+    channel_images_corrected = pybasic.correct_illumination(
+        images_list = channel_images, 
+        flatfield = flatfield, 
+        darkfield = darkfield,
+        )
+        
+    # Covert illum corrected images to uint8 for downstream analysis
+    corrected_images_coverted = np.array(channel_images_corrected)
+    corrected_images_coverted[corrected_images_coverted<0] = 0 # makes the negatives 0
+    corrected_images_coverted = corrected_images_coverted / np.max(corrected_images_coverted) # normalize the data to 0 - 1
+    corrected_images_coverted = 255 * corrected_images_coverted # Scale by 255
+    corrected_images = corrected_images_coverted.astype(np.uint8)
+
+    # Save corrected images with suffix for identification
+    new_channel_save_path = pathlib.Path(save_path, save_channel)
+    
+    for i, image in enumerate(corrected_images):
+        orig_file = image_files[i]
+        orig_file_name = orig_file.split("/")[4]
+        new_filename = f'{new_channel_save_path}/{orig_file_name}_IllumCorrect.tif'
+        skimage.io.imsave(new_filename, image)
+
+
+# ### Implementing helper functions to run illumination correction on each channel for NF1 data
+
+# In[50]:
+
+
+channel_path = pathlib.Path('../PyBaSiC_Pipelines/NF1_Channels')
+channel = "DAPI"
+save_path = pathlib.Path('../PyBaSiC_Pipelines/NF1_Channels')
+save_channel = "DAPI_Corrected"
+# If you want to output the flatfield and darkfield calculations, then put "output_calc=True". If not, leave out from function since it is defaulted to False
+
+run_illum_correct(channel_path, channel, save_path, save_channel)
```

Comment:
I see SO much improvement :) very nice to see.

Let's go one level up now too.

How about something like this:

```python
# Set location of input and output locations
channel_path = pathlib.Path('../PyBaSiC_Pipelines/NF1_Channels')

# What channels to process
channels = ["DAPI", "GFP", "RFP"]

# Perform illumination correction on all channels:
for channel in channels:
    # Folder to save the output channel
    save_channel = f"{channel}_Corrected"

    # Perform illumination correction
    run_illum_correct(
        channel_path=channel_path,
        channel=channel,
        save_path=channel_path,
        save_channel=save_channel
    )
```

Does this make sense? Now you're processing _all_ channels in one command